python/pyspark/sql/dataframe.py

```diff
@@ -1760,13 +1760,39 @@ def toPandas(self):
                       "if using spark.sql.execution.arrow.enable=true"
                 raise ImportError("%s\n%s" % (e.message, msg))
         else:
+            import numpy as np
             dtype = {}
+            nullable_int_columns = set()
+
+            def null_handler(rows, nullable_int_columns):
+                requires_double_precision = set()
+                for row in rows:
+                    row = row.asDict()
+                    for column in nullable_int_columns:
+                        val = row[column]
+                        dt = dtype[column]
+                        if val is None and dt not in (np.float32, np.float64):
+                            dt = np.float64 if column in requires_double_precision else np.float32
+                            dtype[column] = dt
+                        elif val is not None:
+                            if abs(val) > 16777216:  # Max value before np.float32 loses precision.
```

Comment:
Why do we need this?

Reply:
Values above this cannot be represented losslessly as a `np.float32`.

Reply:
I think they are represented as np.float64. I added a test in #19319.